### PR TITLE
Add scalar wave rhs

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add ability to call specific constructor via `make_array`
- Change to the new `make_array` call in `Tensor`
- Add ScalarWave tags
- Add ScalarWave RHS

I'm planning on adding the flux equations in a separate PR that will be dependent on this PR.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: tests/Unit/Utilities/Test_MakeArray.cpp [hunk 1]
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
...
+      make_array<3, MyNonCopyable<std::vector<double>>>(
+          std::vector<double>{2.3, 8.6, 9.7});
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(noncopyable_vectors[0].get() == (std::vector<double>{2.3, 8.6, 9.7}));
+  }
+
+  // Check make_array with a multi-arg constructor
+  const auto vector_array1 = make_array<3, std::vector<int>>(3_st, 9);
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(vector_array1[0] == (std::vector<int>(3, 9)));
```
Comment (wthrowe) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: Again.

File: tests/Unit/Utilities/Test_MakeArray.cpp [hunk 2]
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
...
+  }
+
+  // Check make_array with an initializer list constructor
+  const auto vector_array2 = make_array<3, std::vector<int>>({3, 9, 12, -10});
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(vector_array2[0] == (std::vector<int>{3, 9, 12, -10}));
```
Comment (wthrowe) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: Again.

File: tests/Unit/Utilities/Test_MakeArray.cpp [hunk 3]
```diff
@@ -59,4 +80,24 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
   constexpr const auto array_from_truncated_sequence =
       make_array<int, 3>(std::initializer_list<int>{2, 8, 6, 9, 7});
   CHECK((std::array<int, 3>{{2, 8, 6}}) == array_from_truncated_sequence);
+
+  // Check make_array with a non-copyable type
+  const auto noncopyable_vectors =
+      make_array<3, MyNonCopyable<std::vector<double>>>(
+          std::vector<double>{2.3, 8.6, 9.7});
+  for (size_t i = 0; i < 3; i++) {
+    CHECK(noncopyable_vectors[0].get() == (std::vector<double>{2.3, 8.6, 9.7}));
```
Comment (wthrowe) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: Shouldn't this depend on the loop index somehow?

File: src/Evolution/Systems/ScalarWave/Equations.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+namespace ScalarWave {
+/*!
+ * \brief Compute the RHS of the time derivative of the PDE system
+ *
+ * ScalarWave equation is given by
+ * \f[
+ * \frac{\partial^2 u}{\partial t^2} - \frac{\partial^2 u}{\partial x^2}=0
+ * \f]
+ * See Hesthaven and Warburton, Nodal Discontinuous Galerkin Methods, page
+ * 159, or wikipedia.
```
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/Equations.hpp]: `Wikipedia` (proper noun)

File: src/Utilities/MakeArray.hpp [hunk 5]
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
...
-    // clang-format off
-make_array_impl(
-    T&& value, std::integer_sequence<size_t, indexes...> /* unused */)
-    noexcept(noexcept(std::array<std::decay_t<T>, size>{
-    {(static_cast<void>(indexes), value)..., std::forward<T>(value)}})) {
-  return std::array<std::decay_t<T>, size>{
-      {(static_cast<void>(indexes), value)..., std::forward<T>(value)}};
-}
-// clang-format on
+template <bool SizeZero>
```
Comment (kidder) [inline src/Utilities/MakeArray.hpp]: please update the comment; there is no `indexes` or `value` or `size`

File: src/Utilities/MakeArray.hpp [hunk 6]
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
...
-template <typename T>
-SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, 0> make_array(
-    std::integral_constant<std::size_t, 0> /* unused */,
-    T&& /* unused */) noexcept {
-  return std::array<std::decay_t<T>, 0>{{}};
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Create a `std::array<T, size>{{T(args...), T(args...), ...}}`
+ * \tparam Size the size of the array
+ * \tparam T the type of the element is the array
```
Comment (kidder) [inline src/Utilities/MakeArray.hpp]: is -> in

File: src/Utilities/MakeArray.hpp [hunk 7]
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
...
-/// \cond HIDDEN_SYMBOLS
-/// Construct empty array specialization
-template <typename T>
-SPECTRE_ALWAYS_INLINE constexpr std::array<std::decay_t<T>, 0> make_array(
-    std::integral_constant<std::size_t, 0> /* unused */,
-    T&& /* unused */) noexcept {
-  return std::array<std::decay_t<T>, 0>{{}};
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Create a `std::array<T, size>{{T(args...), T(args...), ...}}`
```
Comment (kidder) [inline src/Utilities/MakeArray.hpp]: size -> Size

File: src/Evolution/Systems/ScalarWave/System.hpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class ScalarWaveSystem.
+
+#pragma once
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Variables.hpp"
```
Comment (wthrowe) [inline src/Evolution/Systems/ScalarWave/System.hpp]: These first two includes are unused.

File: src/Utilities/MakeArray.hpp [hunk 9]
```diff
@@ -27,54 +27,59 @@ namespace MakeArray_detail {
...
-  return MakeArray_detail::make_array_impl<size>(
-      std::forward<T>(value), std::make_index_sequence<size - 1>{});
-}
-// clang-format on
 /*!
  * \ingroup UtilitiesGroup
- * \brief Helper class to initialize a std::array.
- *
- * \tparam size the length of the array
+ * \brief Create a `std::array<std::decay_t<T>, size>{{t, t, ...}}`
```
Comment (kidder) [inline src/Utilities/MakeArray.hpp]: size -> Size

File: tests/Unit/Utilities/Test_MakeArray.cpp (hunk 3 above)
Comment (kidder) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: shouldn't this check all the elements of the array? i.e. `[0]` -> `[i]`

File: tests/Unit/Utilities/Test_MakeArray.cpp (hunk 1 above)
Comment (kidder) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: ditto

File: tests/Unit/Utilities/Test_MakeArray.cpp (hunk 2 above)
Comment (kidder) [inline tests/Unit/Utilities/Test_MakeArray.cpp]: ditto

File: src/Evolution/Systems/ScalarWave/Equations.hpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace ScalarWave {
+/*!
+ * \brief Compute the RHS of the time derivative of the PDE system
+ *
+ * ScalarWave equation is given by
+ * \f[
+ * \frac{\partial^2 u}{\partial t^2} - \frac{\partial^2 u}{\partial x^2}=0
+ * \f]
+ * See Hesthaven and Warburton, Nodal Discontinuous Galerkin Methods, page
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/Equations.hpp]: I think this comment should describe the first-order form of the scalar wave equation

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/ScalarWave/Tags.hpp [hunk 11]
```diff
@@ -0,0 +1,28 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines DataBox tags for scalar wave system
+
+#pragma once
+
+#include "DataStructures/DataBoxTag.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/Tags.hpp]: includes (or forward declarations) for size_t and DataVector

File: src/Evolution/Systems/ScalarWave/System.hpp [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+#include "Evolution/Systems/ScalarWave/Equations.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/System.hpp]: first two not yet used

File: src/Evolution/Systems/ScalarWave/System.hpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
+
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class ScalarWaveSystem.
+
+#pragma once
+
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/System.hpp:9]: include for size_t

File: src/Evolution/Systems/ScalarWave/System.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/ScalarWave/Equations.hpp"
+#include "Evolution/Systems/ScalarWave/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ScalarWave {
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/System.hpp]: does this namespace have a doxygen comment anywhere?

File: src/Evolution/Systems/ScalarWave/Equations.hpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Utilities/Gsl.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/ScalarWave/Equations.hpp]: forward declaration sufficient?

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
While addressing the comment about documenting the ScalarWave namespace I added a three commits. One adds a Doxygen group for evolution systems, another removes the currently unused and non-alphabetical ParallelComponentsGroup, and the last one adds GH to the evolution systems group.

File: src/Utilities/MakeArray.hpp [hunk 16]
```diff
@@ -19,14 +19,15 @@
...
 namespace MakeArray_detail {
-/// We take integer_sequence<indexes...> as a parameter to overload the
-/// function so that there is one for each type T, and one for each number of
-/// indexes. To silence compiler warnings we do a static_cast<void>.
-/// Use the comma operator to expand the parameter pack, then copy in value
-/// (size-1) times. We use std::forward to try and move the last value into
-/// place, rather than undergoing another copy.
-/// Order of evaluation is well-defined for aggregate initialization, so there
-/// is no risk of copy-after-move
+// We handle the zero size and case separately below because both for zero size
```
Comment (kidder) [inline src/Utilities/MakeArray.hpp]: no and

File: docs/GroupDefs.hpp [hunk 17]
```diff
@@ -90,6 +90,11 @@
  * Macros and functions used for handling errors
  */
 
+/*!
+ * \defgroup EvolutionSystemsGroup Evolution Systems
+ * \brief Contains namespace of all the available evolution systems.
```
Comment (kidder) [inline docs/GroupDefs.hpp]: the namespaces

**Review by kidder — CHANGES_REQUESTED** [review]
as far as I am concerned you can squash when making the latest changes

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**